### DISCUSSION TITLE
Relevancy of GatewayClass SupportedVersion Feature

### DISCUSSION THREAD

**Discussion by danehans** [kubernetes-sigs/gateway-api#3494] [General]
PR https://github.com/kubernetes-sigs/gateway-api/pull/3368 adds a conformance test for the GatewayClass [SupportedVersion](https://github.com/kubernetes-sigs/gateway-api/blob/v1.2.1/apis/v1/gatewayclass_types.go#L204-L235) feature. The PR sparked discussion regarding the feature's relevancy, whether CRDs should be watched by controllers, etc. This discussion should be used to continue the conversation instead.

**Comment by danehans** [comment]
@robscott feel free to provide additional context if needed.

**Comment by robscott** [comment]
Thanks for starting this conversation @danehans! In my opinion, the strongest arguments in favor of something like this would be to warn users when:

1) They've installed old version of CRD, potentially with older or newer API versions (v1alpha2, v1) that the controller doesn't support
2) They're using features that the controller doesn't support (potentially due to feature being exclusive to a newer API than the controller supports)
3) They're using a tool to manage CRDs that wants to know if an upgrade is safe by checking CRD status to find the max supported version across controllers.

It's likely that 2 can be solved with the `supportedFeatures` field, it's unclear how we'd resolve 1.

The strongest arguments against something like this would be:

1. Gateway API is backwards compatible, customers shouldn't really care about individual API versions, as long as the features they care about are present
2. Some potential solutions here would be onerous for controllers to support

A caveat to 1 is that Gateway API only offers backwards compatibility guarantees on standard channel, not experimental channel. It could be incredibly useful to understand if an implementation supports v1.1 (BackendTLSPolicy v1alpha3) or v1.0 (BackendTLSPolicy v1alpha2). You could argue that the important detail there is API Version, not bundle/semantic version though.

/cc @howardjohn @youngnick

**Comment by candita** [comment]
As mentioned in the meeting today, it should count for something that we've documented this in the [1.0 CHANGELOG](https://github.com/kubernetes-sigs/gateway-api/blob/main/CHANGELOG/1.0-CHANGELOG.md?plain=1#L130-L131), declared it an Issue in [#2077](https://github.com/kubernetes-sigs/gateway-api/issues/2077), and implemented it in ([#2384](https://github.com/kubernetes-sigs/gateway-api/pull/2384)).

More importantly, there are at least two implementations of `SupportedVersion` already - [Nginx Gateway Fabric](https://github.com/nginxinc/nginx-gateway-fabric/pull/1301) and [Contour](https://github.com/projectcontour/contour/pull/6147)

**Comment by howardjohn** [comment]
For the record I still disagree with putting project documentation into resource statuses (and therefor `SupportedVersions` as a concept), and its not a concept I have seen in any other project's APIs
